v2.1/performance-best-practices-overview.md

```diff
@@ -226,3 +226,88 @@ This query returns the account balances of the customers.
+  same [data ranges](architecture/overview.html).
+- At least some of the transactions write or modify the data.
+
+A set of transactions that all contend on the same keys will be
+limited in performance to the maximum processing speed of a single
+node (limited horizontal scalability). Non-contended transactions are
+not affected in this way.
+
+There are two levels of contention:
+
+- Transactions that operate on the same range but different index key
+  values will be limited by the overall hardware capacity of a single
+  node (the range lease holder).
+
+- Transactions that operate on the same index key values
+  (specifically, that operate on the same [column
+  family](column-families.html) for a given index key) will be more
+  strictly serialized to obey transaction isolation semantics.
+
+Transaction contention can also increase the rate of transaction
+restarts, and thus make the proper implementation of [client-side
+transaction
+retries](transactions.html#client-side-transaction-retries) more
+critical.
+
+To avoid contention, multiple strategies can be applied:
+
+- Use index key values with a more random distribution of values, so
+  that transactions over different rows are more likely to operate on
+  separate data ranges. See the [SQL FAQs](sql-faqs.html) on row IDs for
+  suggestions.
+
+- Make transactions smaller, so that each transaction has less work to
+  do. In particular, avoid multiple client-server exchanges per
+  transaction. For example, use [common table
+  expressions](common-table-expressions.html) to group multiple
+  [`SELECT`](select-clause.html) and
+  [`INSERT`](insert.html)/[`UPDATE`](update.html)/[`DELETE`](delete.html)/[`UPSERT`](upsert.html)
+  clauses together in a single SQL statement.
+
+- When replacing values in a row, use [`UPSERT`](upsert.html) and
+  specify values for all columns in the inserted rows. This will
+  usually have the best performance under contention, compared to
+  combinations of [`SELECT`](select-clause.html),
+  [`INSERT`](insert.html) and [`UPDATE`](update.html).
+
+- Increase
+  [normalization](https://en.wikipedia.org/wiki/Database_normalization)
+  of the data to place parts of the same records that are modified by
+  different transactions in different tables. Note however that this
+  is a double-edged sword, because denormalization can also increase
+  performance by creating multiple copies of often-referenced data in
+  separate ranges.
+
+- If the application strictly requires operating on very few different
+  index key values, consider using [`ALTER ... SPLIT
+  AT`](split-at.html) so that each index key value can be served by
+  a separate group of nodes in the cluster.
+
+It is always best to avoid contention as much as possible via the
+design of the schema and application. However, sometimes contention is
+unavoidable. To maximize performance in the presence of contention,
+you'll need to maximize the performance of a single range.
```

Comment:
Is this paragraph introducing the following bulleted list? If so, use a semicolon instead of a period at the end of the last sentence.